dojo/tools/gitlab_dast/parser.py

```diff
@@ -0,0 +1,186 @@
+
+    def get_scan_types(self):
+        return ["GitLab DAST Report"]
+
+    def get_label_for_scan_types(self, scan_type):
+        return "GitLab DAST Report"
+
+    def get_description_for_scan_types(self, scan_type):
+        return "GitLab DAST Report in JSON format (option --json)."
+
+    # turning a json file to string
+    def parse_json(self, file):
+        data = file.read()
+        try:
+            tree = json.loads(str(data, 'utf-8'))
+        except:
+            tree = json.loads(data)
+
+        return tree
+
+    def get_items(self, tree, test):
+        items = {}
+
+        # iterating through each vulnerability
+        for node in tree['vulnerabilities']:
+            item = get_item(node, test)
+            if item:
+                items[item.unique_id_from_tool] = item
+
+        return list(items.values())
+
+    def get_findings(self, file, test):
+        if file is None:
+            return None
+
+        tree = self.parse_json(file)
+        if tree:
+            return self.get_items(tree, test)
+
+    def convert_severity(self, num_severity):
+        """Convert severity value"""
+        if num_severity >= -10:
+            return "Low"
+        elif -11 >= num_severity > -26:
+            return "Medium"
+        elif num_severity <= -26:
+            return "High"
+        else:
+            return "Info"
+
+
+# iterating through properties of each vulnerability
+def get_item(vuln, test):
+
+    if vuln["category"] != "dast":
+        return None
+
+    # scanner_confidence
+    scanner_confidence = get_confidence_numeric(vuln["confidence"])
+
+    # id
+    if "id" in vuln:
+        unique_id_from_tool = vuln["id"]
+    else:  # deprecated
+        unique_id_from_tool = vuln["cve"]
+
+    # title
+    if "name" in vuln:
+        title = vuln["name"]
+    # fallback to using id as a title
+    else:
+        title = unique_id_from_tool
+
+    # description
+    description = f"Scanner: {vuln['scanner']['name']}\n"
+    if "message" in vuln:
+        description += f"{vuln['message']}\n"
+    elif "description" in vuln:
+        description += f"{vuln['description']}\n"
+
+    # date
+    if "discovered_at" in vuln:
+        temp = vuln["discovered_at"][:-4]
+        date = datetime.strptime(temp, "%Y-%m-%dT%H:%M:%S")
+    else:
+        date = None
+
+    # endpoint
+    location = vuln["location"]
+    if "hostname" in location and "path" in location:
+        url = f"{location['hostname']}{location['path']}"
+
+        o = urlparse(url)
+        protocol = o.scheme
+
+        port = 80
+        if protocol == "https":
+            port = 443
+        if o.port is not None:
+            port = o.port
+
+        host = o.netloc
+        query = o.query
+        fragment = o.fragment
+        path = o.path
+
+        endpoint = Endpoint(
+            protocol=protocol,
+            host=host,
+            port=port,
+            query=query,
+            fragment=fragment,
+            path=path
+        )
```

Comment:
Since we merged #4473 , we don't need to parse and analyze URLs anymore in parsers.
You should switch to `Endpoint.from_url()` method.

Reply:
Sorry for asking this since I'm still pretty inexperienced, but does this imply that I can instead do `endpoint = Endpoint.from_uri(url)`?

(Note that the method provided in models.py is `from_uri` not `from_url`

Reply:
Ok nvm, I think I got it :)

Reply:
This is how I currently create my endpoint:
```# endpoint
    location = vuln["location"]
    if "hostname" in location and "path" in location:
        url = f"{location['hostname']}{location['path']}"
        endpoint = Endpoint.from_uri(url)
    else:
        endpoint = None
```
This is the result of the test for the endpoint:
```======================================================================
FAIL: test_parse_file_with_multiple_vuln_has_multiple_findings (dojo.unittests.tools.test_gitlab_dast_parser.TestGitlabDastParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/dojo/unittests/tools/test_gitlab_dast_parser.py", line 107, in test_parse_file_with_multiple_vuln_has_multiple_findings
    self.assertEqual("http", endpoint.protocol)
AssertionError: 'http' != None

======================================================================
FAIL: test_parse_file_with_one_vuln_has_one_finding (dojo.unittests.tools.test_gitlab_dast_parser.TestGitlabDastParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/dojo/unittests/tools/test_gitlab_dast_parser.py", line 67, in test_parse_file_with_one_vuln_has_one_finding
    self.assertEqual("http", endpoint.protocol)
AssertionError: 'http' != None
```
Notice that in this case, endpoint.protocol == None, which did not happen before when I was doing it the old way. Is this a bug or is there other ways for me to obtain the protocol of the endpoint?

Reply:
use the method `Endpoint.from_uri()` it will make the code simpler and less bug prone

Reply:
Ok, done